lib/internal/fs/promises.js

```diff
@@ -250,6 +251,27 @@ class FileHandle extends EventEmitterMixin(JSTransferable) {
   }
 }
 
+async function handleFdClose(fileOpPromise, closeFunc) {
+  let opError;
+  try {
+    return await fileOpPromise;
+  } catch (err) {
+    opError = err;
+    throw err;
+  } finally {
+    try {
+      await closeFunc();
+    } catch (closeError) {
+      if (!opError) {
+        // eslint-disable-next-line no-unsafe-finally
+        throw closeError;
+      }
+      // eslint-disable-next-line no-unsafe-finally
+      throw aggregateTwoErrors(closeError, opError);
+    }
+  }
```

Comment:
I wonder if it wouldn't be more readable _without_ async/await syntax. wdyt?
```suggestion
  return PromisePrototypeThen(
      fileOpPromise,
      (result) => PromisePrototypeThen(closeFunc(), () => result),
      (opError) =>
        PromisePrototypeThen(
          closeFunc(),
          () => PromiseReject(opError),
          (closeError) => PromiseReject(aggregateTwoErrors(closeError, opError))
        )
    );
```

Reply:
Yes, I think that the `cb` version was way better with regards to readability. I'll try your suggestion and play around with it.

Reply:
@aduh95 Could using `PromisePrototypeThen` here cause any performance issue? I'm wondering because https://github.com/nodejs/node/pull/38246#discussion_r614746496 mentions about the `*Prototype*` primordials (though only the ones related to functions and arrays). Just trying make sure that `PromisePrototype*` does not fall into the list.

Reply:
@RaisinTen I highly doubt so, but I've been wrong before :)

Reply:
At least for readfile it looks like a wash
```
                                                              confidence improvement accuracy (*)   (**)  (***)
fs/readfile-promises.js concurrent=1 len=1024 duration=5                      2.18 %       ±3.39% ±4.56% ±6.04%
fs/readfile-promises.js concurrent=1 len=16777216 duration=5                  1.18 %       ±2.41% ±3.20% ±4.17%
fs/readfile-promises.js concurrent=10 len=1024 duration=5                     0.28 %       ±1.38% ±1.84% ±2.41%
fs/readfile-promises.js concurrent=10 len=16777216 duration=5                 0.45 %       ±1.18% ±1.57% ±2.05%
```

Reply:
If it's of any interest, this is the non-primordials version vs primordials (both have the same logic, without async/await):
```
                                                              confidence improvement accuracy (*)   (**)  (***)
fs/readfile-promises.js concurrent=1 len=1024 duration=5                     -0.02 %       ±1.48% ±1.96% ±2.56%
fs/readfile-promises.js concurrent=1 len=16777216 duration=5                 -0.21 %       ±1.49% ±1.98% ±2.59%
fs/readfile-promises.js concurrent=10 len=1024 duration=5                     0.40 %       ±0.99% ±1.32% ±1.72%
fs/readfile-promises.js concurrent=10 len=16777216 duration=5                 0.18 %       ±1.30% ±1.73% ±2.25%
```